internal/uvm/virtual_device.go

```diff
@@ -15,54 +15,106 @@ import (
 // belongs to.
 type VPCIDevice struct {
 	vm *UtilityVM
-	ID string
+	// VMBusGUID is the GUID associated with the instance of a new vmbus child
+	// device, vpci, from which the target assigned device will be attached to
+	// in the UVM. This is used to locate the assigned devices in the UVM's view.
+	VMBusGUID string
+	// deviceInstanceID is the instance ID of the device on the host
+	deviceInstanceID string
+	refCount         uint32
 }
 
 // Release frees the resources of the corresponding vpci device
 func (vpci *VPCIDevice) Release(ctx context.Context) error {
-	if err := vpci.vm.RemoveDevice(ctx, vpci.ID); err != nil {
+	if err := vpci.vm.RemoveDeviceFromUVM(ctx, vpci); err != nil {
 		log.G(ctx).WithError(err).Warn("failed to remove VPCI device")
 		return err
 	}
 	return nil
 }
 
-// AssignDevice assigns a new vpci device to the uvm
-func (uvm *UtilityVM) AssignDevice(ctx context.Context, device hcsschema.VirtualPciDevice) (*VPCIDevice, error) {
+// AssignDeviceToUVM assigns a vpci device to the uvm
+// if the device already exists, the stored VPCIDevice's ref count is increased
+// and the VPCIDevice is returned.
+// Otherwise, a new request is made to assign the target device indicated by the deviceID
+// onto the UVM. A new VPCIDevice entry is made on the UVM and the VPCIDevice is returned
+// to the caller
+func (uvm *UtilityVM) AssignDeviceToUVM(ctx context.Context, deviceID string) (*VPCIDevice, error) {
 	guid, err := guid.NewV4()
 	if err != nil {
 		return nil, err
 	}
-	id := guid.String()
+	vmBusGUID := guid.String()
 
 	uvm.m.Lock()
 	defer uvm.m.Unlock()
-	if err := uvm.modify(ctx, &hcsschema.ModifySettingRequest{
-		ResourcePath: fmt.Sprintf(virtualPciResourceFormat, id),
+
+	existingVPCIDevice := uvm.vpciDevices[deviceID]
+	if existingVPCIDevice != nil {
+		existingVPCIDevice.refCount++
+		return existingVPCIDevice, nil
+	}
+
+	targetDevice := hcsschema.VirtualPciDevice{
+		Functions: []hcsschema.VirtualPciFunction{
+			{
+				DeviceInstancePath: deviceID,
+			},
+		},
+	}
+
+	request := &hcsschema.ModifySettingRequest{
+		ResourcePath: fmt.Sprintf(virtualPciResourceFormat, vmBusGUID),
 		RequestType:  requesttype.Add,
-		Settings:     device,
-		GuestRequest: guestrequest.GuestRequest{
+		Settings:     targetDevice}
+
+	// we do not currently support assigned devices on windows,
+	// so no guest side work necessary for windows
```

Comment:
Maybe I am misreading this, but if we don't support calling this function at all for a Windows UVM, can we move the check to the top of the function and return an error if it's WCOW?

Reply:
I added a case at the beginning of the function, but am keeping this check since we will be supporting device assignment in wcow very soon :P